C/C++ guide/tutorial (#2166)
emacs-lsp/lsp-mode: docs/blog/2020/CPP-guide.md

---

> disqus: emacs-lsp, C++/C

Reviewer: I think this should be our disqus username
```suggestion
disqus: emacs-lsp
```

---

> # Configuring `Emacs` as `C/C++` IDE and for `Emacs` core development

Reviewer: I'd reduce the title for some for attractive like:
```suggestion
# Configuring `Emacs` a `C/C++` IDE
```
And maybe mention about emacs core development somewhere bellow

In this guide, I will help you to setup an efficient working `C/C++`
environment. The guide is mainly focused on `lsp-mode`/`dap-mode` and the

> features t. We will use `GNU Emacs` as a project in order to illustrate each of

Reviewer: ```suggestion
features that they have. We will use `GNU Emacs` as a project in order to illustrate each of
```

the steps to get `C/C++` project up and running.

Some of the main features that we can expect from C++ IDE:

* Code completion
* Real-time syntax checking
* Code navigation (references/definitions/implementations/symbol search)
* Visual IDE debugger

> 

Reviewer: Maybe mention other lsp main features
```suggestion

For all `lsp-mode` features, check [here](https://emacs-lsp.github.io/lsp-mode/page/main-features/)
```

itself when a certain feature is supported by the server and when the
corresponding `Emacs` package is installed. In addition to that, `lsp-mode`
ships with [which-key](https://github.com/justbur/emacs-which-key "which key url") integration to speedup functionality discoverability.
Keybindings are bound are available under `s-l` (Super-l), e. g. `s-l t t` will
show documentation at point. You may change the default prefix by setting
`lsp-keymap-prefix`.

![select project root](images/which-key.png "Select project root")

### Mouse support
Yeah, `lsp-mode` supports mouse!

![mouse support](images/mouse.png "Mouse support")

### Refactoring/Code actions

- `lsp-rename` (`s-l r r`) - rename symbol/function at point.
- `lsp-execute-code-action` (`s-l a a`) - `clangd` is able to auto-correct some of the errors.

### Navigation
`lsp-mode` has integration with `xref` core package and in addition it has

- `xref-find-definitions`(`M-.` and `s-l g g`) - find definition(s) at point
- `xref-find-references`(`s-l g r`) - find references to the symbol at point
![References](images/images/references-helm.png "References helm")

> - `lsp-find-implementation`(`s-l g i`) - find implementation

Author: Yes, I will delete the reference.

in the minibuffer. `lsp-mode` is showing only the one line signature info so if
you want to see the full documentation at point you can use `lsp-describe-thing-at-point`.

![hover](images/hover-info.png)

In addition to that you can also see signature help after pressing `C-M-SPC` or
after pressing trigger char like `(`. When there are more than one signature you
may browse them via `M-n/M-p`

![signature info](images/signature-help.png)

### Diagnostics
For on-the-fly errors `lsp-mode` is using `flycheck` (`flymake` is also
supported). It configured automatically. In addition to standard `flycheck`
features `lsp-mode` provides project level errors browser via
`lsp-treemacs-errors-list`.

![errors list](images/errors-list.png)

## Debugging

Debugging in `lsp-mode` is implemented via [dap-mode](https://github.com/emacs-lsp/dap-mode) package. `dap-mode`
front-end of the sister [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/). It is again client server
like `Language Sever Protocol`. `dap-mode` provides all of the traditional
debugger features - breakpoints(conditions, hit count, etc), threads, locals,

### Installation

> `dap-mode` provides installation commands for most of the debug adapters. This

Reviewer 2: ```suggestion
`dap-mode` provides installation commands for most of the debug adapters. In this
```

> tutorial we will use [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter

Reviewer 2: ```suggestion
tutorial we will use the [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter
```

do `M-x dap-cpptools-setup`.

> _Note:_ you should have `gbd` on path.

Reviewer 2: ```suggestion
_Note:_ you should have `gdb` on path.
```

being deprecated. For this guide we will be using `clangd` but most of the steps here
apply to `ccls` as well. `lsp-mode` is pre-configured and it will be able to find
the language server automatically when it is installed on the host machine and
it is present on the path. In case `clangd` is not present on the path
`clangd` can be installed following their [Getting Started](https://clangd.llvm.org/installation.html "Getting Started")
instructions.

_Note:_ for `Debian` based system use https://apt.llvm.org/

## Project setup

To understand your source code, `clangd` needs to know your build flags. (This is
just a fact of life in C++, source files are not self-contained).

By default, `clangd` will assume your code is built as `clang` `some_file.cc`, and
you’ll probably get spurious errors about missing `#included` files, etc. There
are a couple of ways to fix this. You may read about them in [installation](https://clangd.llvm.org/installation.html "clangd installation")

In order to generate `compile_commands.json` for `Emacs` itself I have found
that [Bear](https://github.com/rizsotto/Bear) works fine.

``` bash

> git clone https://github.com/rizsotto/Bear

Reviewer 3: `Bear` can be installed via `apt` too `sudo apt install bear`

Author: thanks, updated.

production-ready language servers for `C/C++` and `Objective-C`,
[ccls](https://github.com/MaskRay/ccls) and
[clangd](https://clangd.llvm.org/installation.html), with `ccls`'s precursor
`cquery` being deprecated. used by `Emacs` as provider for language specific
features(completion, diagnostics, navigation, etc). For some of the language
servers `lsp-mode` ships with automatic installation scripts but there is no
such script for installing `C/C++` language server due to their overall
complexity. As of end of 2020 there are 2 `C/C++` language servers - `ccls` and
`clangd` with `cquery` being deprecated. For this guide we will be using
`clangd` but most of the steps here apply to `ccls` as well. `lsp-mode` is
pre-configured and it will be able to find the language server automatically
when it is installed on the host machine and it is present on the path. In case
`clangd` is not present on the path `clangd` can be installed following their
[Getting Started](https://clangd.llvm.org/installation.html "Getting Started")
instructions.

_Note:_ for `Debian` based system use https://apt.llvm.org/

## Project setup

> To understand your source code, `clangd` needs to know your build flags. (This is

Author: I will apply that, but I realized that I have copied that section directly from clangd docs(but I already applied few comments on it).

just a fact of life in C++, source files are not self-contained).

By default, `clangd` will assume your code is built as `clang` `some_file.cc`, and
you’ll probably get spurious errors about missing `#included` files, etc. There
are a couple of ways to fix this. You may read about them in [installation](https://clangd.llvm.org/installation.html "clangd installation")

> In order to generate `compile_commands.json` for `Emacs` itself I have found

Author: It is mentioned in the linked Clangd resource - https://clangd.llvm.org/installation.html , I can copy their section?

We might consider providing also a "Hello world" example and also whatever you C++ guys think is relevant.

> In order to generate `compile_commands.json` for `Emacs` itself I have found

Reviewer 4: https://sarcasm.github.io/notes/dev/compilation-database.html discusses various ways to get compile_commands.json files.

(package-initialize)

(setq package-selected-packages '(lsp-mode yasnippet lsp-treemacs helm-lsp
    projectile hydra flycheck company avy which-key helm-xref dap-mode))

(when (cl-find-if-not #'package-installed-p package-selected-packages)
  (package-refresh-contents)
  (mapc #'package-install package-selected-packages))

;; sample `helm' configuration use https://github.com/emacs-helm/helm/ for details
(helm-mode)
(require 'helm-xref)
(define-key global-map [remap find-file] #'helm-find-files)
(define-key global-map [remap execute-extended-command] #'helm-M-x)
(define-key global-map [remap switch-to-buffer] #'helm-mini)

(which-key-mode)
(add-hook 'c-mode-hook 'lsp)
(add-hook 'cpp-mode-hook 'lsp)

(setq gc-cons-threshold (* 100 1024 1024)
      read-process-output-max (* 1024 1024)
      treemacs-space-between-root-nodes nil
      company-idle-delay 0.0
      company-minimum-prefix-length 1
      lsp-idle-delay 0.1 ;; clangd is fast
      ;; be more ide-ish
      lsp-headerline-breadcrumb-enable t)

(with-eval-after-load 'lsp-mode
  (add-hook 'lsp-mode-hook #'lsp-enable-which-key-integration)
  (require 'dap-cpptools)
  (yas-global-mode))
```

> # Feature overview.

Reviewer 5: ```suggestion
# Feature overview
```

And now it is time to start hacking on `Emacs` core!

```
emacs ~/Sources/emacs/src/editfns.c
```

Now, if everything was successful you will be asked to select a project root.

![select project root](images/import-project.png "Select project root")

_Note:_ the project root is needed by the language server in order to know where to
start the project discovery from. Typically, this is the git repository root but since exceptions to this rule have caused us a lot of trouble in the past (monorepos come to mind), `lsp-mode` by default asks the user to manually confirm the project root the first time a project is opened.
mono-repos it might be not. `lsp-mode` by default will ask the user to confirm
the project root (automatic root selection has caused a lot of trouble for us in

> the past).

Author: I prefer not to mention that option here. In the beginning, it was part of popular guides and a lot of users put it in their config which caused a lot of issues/complaints.

supported). It is configured automatically. In addition to standard `flycheck`
features `lsp-mode` provides a project-wide error list via
`lsp-treemacs-errors-list`.

![errors list](images/errors-list.png)

## Debugging

Just like `lsp-mode` provides editing features through the editor-independent [Language Server Protocol](https://microsoft.github.io/language-server-protocol/), its sister package `dap-mode` provides debugging features through the editor-independent [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/) and language-specific debugging servers called _debug adapters_.
front-end of the sister [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/). It is again client server
like `Language Sever Protocol`. `dap-mode` provides all of the traditional
debugger features - breakpoints(conditions, hit count, etc), threads, locals,

### Installation

`dap-mode` provides installation commands for most of the debug adapters. In this
tutorial we will use the [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter
do `M-x dap-cpptools-setup`.

> _Note:_ you should have `gdb` on path.

Reviewer 5: For macOS users, `lldb` is more popular than `gdb` (which is not available in the base system). As the debug adapter supports both I can expand the guide later to show how it works with `lldb`.

Author: Yes. I was about to suggest someone using macOS(or other popular OS-es) to provide the macos specific steps.

in project root and `dap-mode` will pick it up.

``` json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug Emacs",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/src/emacs",
            "args": ["-q"],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}",
            "environment": [],
            "externalConsole": true,
            "MIMode": "gdb"
        }
    ]
}
```

At this point, `dap-mode` does not provide the list of all available properties
for each configuration. The full list can be found in `package.json` in
`/.emacs.d/.extension/vscode/cpptools/extension/package.json` when `dap-mode` is
using `VScode` package or debug adapter docs when not.

### Debugging

Start debugging via `M-x dap-debug` and select `Debug Emacs` from the list.
![select debug template](images/select-template.png "Select debug template").

After pressing `RET` a fresh instance of `Emacs` will pop up. Then do `M-x
helm-lsp-workspace-symbol` and type `message3`. Go to that function and place a
breakpoint. You could do that by clicking in the fringe or by doing `M-x

> dap-breakpoint-toggle`. Switch to debugged instance and do

Reviewer 5: ```suggestion
dap-breakpoint-toggle`. Switch to the debugged instance and do
```
